### PR TITLE
HW-30 done

### PR DESCRIPTION
**Pull request by verovkin**
done :)

### CODE REVIEW

File: 30/task2.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import requests
+
+city = input("Enter location: ")
+
+CITY_API = f"https://geocoding-api.open-meteo.com/v1/search?name={city}&count=1"
+city_res = requests.get(CITY_API)
+if city_res.status_code == 200:
+    city_res_data = city_res.json()
+    latitude = city_res_data.get("results")[0]["latitude"]
```
Comment (klymenok) [inline 30/task2.py]: You would get an error here if user entered a non-existing city

**Review by klymenok — CHANGES_REQUESTED** [review]

**Comment by verovkin** [comment]
Rewrote code in OOP, interesting is it written in such way :)

File: 30/task2.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            weather_res = requests.get(WEATHER_FORECAST_API_URL, params=weather_api_params)
+            if weather_res.status_code == 200:
+                weather_res_data = weather_res.json().get("current_weather")
+                print(f"The weather in {self.name} is - temperatute {weather_res_data['temperature']}°C and the wind speed is {weather_res_data['windspeed']} km/h")
+            else:
+                print("Error getting a weather")
+                exit()
+
+
+city = City()
```
Comment (klymenok) [inline 30/task2.py]: Good try, but to improve your OOP approach, I would suggest the following:
1. Use objects and methods and something that you can use and reuse
2. Initialize object's properties in constructor and raise error if something is wrong. You can take `name` as a parameter in the constructor and then call `protected` method `set_coordinates`, to init attributes `self.lat` and `self.lon`
3. Make methods smaller and responsible only for some atomic task

In the end you should get something like this:
```
class City:
    def __init__(self, name):
        self.name = name
        self._set_coordinates()

    def _set_coordinates(self):
        # some code here
        if not lat or not lon:
            raise ValueError('No such city')
        self.lat = lat
        self.lon = lon

    def print_weather(self):
        # some code here
        if weather.status_code == 200:
            print(...)
        else:
            print(...)

city_name = input()
try:
    city = City(city_name)
except ValueError as e:
    print(e)
else:
    city.print_weather()
```

If you want to go further, you can create a separate class to work with weather API and unload your city class. You can keep in city only city-related information:
```
class City:
    def __init__(self, name):
        self.name = name
        self._set_coordinates()

    def _set_coordinates(self):
        self.lat, self.lon = WeatherService.get_coordinates(self.name)

    def print_weather(self):
        weather_data = WeatherService.get_current_weather(self.lat, self.lon)
        if weather_data:
            print(...)
        else:
            print(...)

**Review by klymenok — CHANGES_REQUESTED** [review]

File: 30/task2.py (hunk 2 above)
Comment (verovkin) [reply 30/task2.py]: why not include city name enter in City class __init__?

that way we can hide this in the class:
`try:
    city = City(city_name)
except ValueError as e:
    print(e)
else:
    city.print_weather()`

Comment (verovkin) [reply 30/task2.py]: for user it will be easier to use library without try-except

Comment (verovkin) [reply 30/task2.py]: incapsulation, что ли )

Comment (verovkin) [reply 30/task2.py]: I meen we can handle exceptions inside City class. if no city - ask user to enter again

Comment (klymenok) [reply 30/task2.py]: It depends on what your `City` will do. Your suggestion also makes sense. You can check city existing with `WeatherService` on init

File: 30/task2.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        else:  # if coordinates API is not available - quit
+            raise ValueError("Geo coordinate service is not working")
+
+    @classmethod
+    def get_current_weather(self, lat, lon):
+        # weather API
+        WEATHER_FORECAST_API_URL = "https://api.open-meteo.com/v1/forecast"
+        weather_api_params = {
+            "latitude": lon,
```
Comment (klymenok) [inline 30/task2.py]: Lat and Lon are switched here and the result is not for the city entered by a user

**Review by klymenok — CHANGES_REQUESTED** [review]

**Comment by verovkin** [comment]
In this code I use class methods of WeatherService, so, as I understood - no instance is created and no __init__ method is called?

**Comment by klymenok** [comment]
> In this code I use class methods of WeatherService, so, as I understood - no instance is created and no **init** method is called?

Correct. And, you can also implement those methods as `static` since you don't need a reference to your class.

Regular methods - first argument `self` keeps an instance of a class
Class methods - first argument `cls` keeps a class itself
Static methods - arguments are simple arguments, a method has no access either to class or object

**Review by klymenok — APPROVED** [review]
Looks good!

**Pull request merged**